### DISCUSSION TITLE
how to specify --native-file and --cross-file for subprojectcs?

### DISCUSSION THREAD

**Discussion by stsp** [mesonbuild/meson#12602] [Q&A] [answered]
My project consists of multiple subprojects.
Each has its own toolchain description files.
When I configure them by hands, things work.
But when I configure from a top-level project,
how can I specify the needed per-subproject
files?

**Comment by xclaesse** [comment]
That's not possible, the toolchain is global and not per subproject. You have to build them separately.

**Reply by stsp** [reply]
Should I make an RFE ticket about that?
I in principle can use the external
"configure" wrapper. I even intended to.
But there appears a few things, namely
a pkgconf generator, that should better
be done in a top-level project, as it needs
to peek into the sub-projects for various
info.
So now I would prefer, if possible, to avoid
having the artificial configure wrapper.

**Reply by eli-schwartz** [reply]
Why should subprojects be compiled with a different compiler than the top-level project?

**Reply by stsp** [reply]
Top-level project is just a wrapper that
I want to use to generate pkgconf. It
doesn't need any compiler.
The problem with subprojects is that
some of them need a host-mode toolchain,
and some - cross-mode. Even when building
for host, some of them need cross-compilation.

But suppose, things are a bit simpler.
Suppose I can merge the description into
a single file. Even in that case I still don't
know what to do: have small description
files per sub-projects, and a concatenated
one in a top-level project? That still looks
like an unneeded duplication then.

In reality, however, I need to pass
--native-file to some, and --cros-file to
others, so I think merging strategy is
not helping anyway.

**Reply by eli-schwartz** [reply, marked as answer]
You can specify both native files and cross files for the top level project. It will mean that all subprojects *also* have both a native file and a cross file.

Subprojects that use cross targets will use the cross compiler, subprojects that use native targets will use the native compiler. What meson doesn't allow is using multiple native compilers in a single build (or multiple cross compilers, although that one is something we should implement as a new feature for cases where you are targeting multiple cross architectures. For example, QEMU needs this.)

**Reply by stsp** [reply]
> You can specify both native files and cross files for the top level project. It will mean that
> all subprojects also have both a native file and a cross file.

But this I already tried, and it didn't work
for me. Configure in top-level ends with this:
```
  User defined options
    Cross files : ../fdpp/kernel/toolchain.ini
    Native files: ../fdpp/toolchain.ini
```
So far so good.
The content of the native file is:
```
[binaries]
cpp = 'clang++'
```
Yes, its that simple.
Cross-file is more complicated, but lets
see if the native one worked:
```
Executing subproject libfdpp 

libfdpp| Project name: libfdpp
libfdpp| Project version: 0.1
libfdpp| C compiler for the host machine: ccache cc (gcc 13.2.0 "cc (Ubuntu 13.2.0-4ubuntu3) 13.2.0")
libfdpp| C linker for the host machine: cc ld.bfd 2.41
libfdpp| C++ compiler for the host machine: ccache c++ (gcc 13.2.0 "c++ (Ubuntu 13.2.0-4ubuntu3) 13.2.0")
```

So as we can see, even though toplevel
project got the toolchain description file,
it didn't switch the subproject to use that.
Now if you say its supposed to work, what
am I doing wrongly?

**Reply by stsp** [reply]
Hmm, in fact, you are right!
The pit-fall was that in the top-level
project the needed language should
be listed, or otherwise the propagation
doesn't happen.

**Comment by ericwomer** [comment]
The only real clean way to do this is to wrap your build setup in a root Makefile or some other script to call all the other meson.build files per project.
